icu4c/source/test/intltest/messageformat2test_builtin.cpp

```diff
@@ -0,0 +1,181 @@
+
+    test = testBuilder.setPattern("Testing date formatting: {$date :date style=short}.")
+                                .setExpected("Testing date formatting: 11/23/22.")
+                                .setDateArgument(date, TEST_DATE)
+                                .build();
+    TestUtils::runTestCase(*this, test, errorCode);
+
+    test = testBuilder.setPattern("Testing date formatting: {$date :time style=long}.")
+                                .setExpected(CharsToUnicodeString("Testing date formatting: 7:42:37\\u202FPM PST."))
+                                .setDateArgument(date, TEST_DATE)
+                                .build();
+    TestUtils::runTestCase(*this, test, errorCode);
+
+    test = testBuilder.setPattern("Testing date formatting: {$date :time style=medium}.")
+                                .setExpected(CharsToUnicodeString("Testing date formatting: 7:42:37\\u202FPM."))
+                                .setDateArgument(date, TEST_DATE)
+                                .build();
+    TestUtils::runTestCase(*this, test, errorCode);
+
+    test = testBuilder.setPattern("Testing date formatting: {$date :time style=short}.")
+                                .setExpected(CharsToUnicodeString("Testing date formatting: 7:42\\u202FPM."))
+                                .setDateArgument(date, TEST_DATE)
+                                .build();
+    TestUtils::runTestCase(*this, test, errorCode);
+
+    // Error cases
+    // Number as argument
+    test = testBuilder.setPattern(".local $num = {|42| :number}\n\
+                                              {{Testing date formatting: {$num :datetime}}}")
+                                .clearArguments()
+                                .setExpected("Testing date formatting: {|42|}")
+                                .setExpectedError(U_OPERAND_MISMATCH_ERROR)
+                                .build();
+    TestUtils::runTestCase(*this, test, errorCode);
+    // Literal string as argument
+    test = testBuilder.setPattern("Testing date formatting: {|horse| :datetime}")
+                                .setExpected("Testing date formatting: {|horse|}")
+                                .setExpectedError(U_OPERAND_MISMATCH_ERROR)
+                                .build();
+
+    TestUtils::runTestCase(*this, test, errorCode);
+
+}
+
+void TestMessageFormat2::testNumbers(IcuTestErrorCode& errorCode) {
+    CHECK_ERROR(errorCode);
+
+    double value = 1234567890.97531;
+    UnicodeString val = "val";
+
+    TestCase::Builder testBuilder;
+    testBuilder.setName("testNumbers");
+
+    // Literals
+    TestCase test = testBuilder.setPattern("From literal: {123456789 :number}!")
+        .setArgument(val, value)
+        .setExpected("From literal: 123.456.789!")
+        .setLocale(Locale("ro"))
+        .build();
+    TestUtils::runTestCase(*this, test, errorCode);
+
+    test = testBuilder.setPattern("From literal: {|123456789,531| :number}!")
```

Comment:
Good point -- this is an actual bug. According to the spec, it should be an error ( https://github.com/unicode-org/message-format-wg/blob/main/spec/registry.md#number-operands ). I wrote this code before that section of the spec existed, and didn't notice it was wrong because there weren't spec tests that would contradict what the expected output was here.

I fixed this in [cecdf30](https://github.com/unicode-org/icu/pull/2579/commits/cecdf30f32fd661e64cb2a6931562579af794a59) (and noted in comments there that it's not the best solution, but it's what I had time for.)